File: jimmer-ddd-java-sample/domain/src/main/java/org/morecup/jimmerddd/java/sample/domain/goods/RenameGoodsNameWhenOrderProductRename.java
package org.morecup.jimmerddd.java.sample.domain.goods;

import org.morecup.jimmerddd.java.function.MI;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
public class RenameGoodsNameWhenOrderProductRename {
    private final GoodsRepository goodsRepository;

    public RenameGoodsNameWhenOrderProductRename(GoodsRepository goodsRepository) {
        this.goodsRepository = goodsRepository;
    }

    @EventListener
    public void onEvent(RenameEvent event) {
        Goods goods = goodsRepository.findByIdOrErr(event.getGoodsId(), MI.of(GoodsImpl::rename));
        GoodsImpl.proxy.execAndSave(goods, impl -> {
            impl.rename(event.getNewName());
            return null;
        });
    }
}

File: jimmer-ddd-java-sample/domain/src/main/java/org/morecup/jimmerddd/java/sample/domain/goods/RenameGoodsNameWhenOrderProductRename.kt
package org.morecup.jimmerddd.kotlin.sample.domain.goods

import org.morecup.jimmerddd.kotlin.sample.domain.order.RenameEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
class RenameGoodsNameWhenOrderProductRename(
    val goodsRepository: GoodsRepository
) {
    @EventListener
    fun onEvent(event: RenameEvent) {
        val goods: Goods = goodsRepository.findByIdOrErr(event.goodsId,GoodsImpl::rename)
        goodsAggregateProxy.execAndSave(goods) { draft ->
            GoodsImpl(draft).rename(event.newName)
        }
    }
}